### DISCUSSION TITLE
Using symlinks in onedir mode

### DISCUSSION THREAD

**Discussion by zevlee** [pyinstaller/pyinstaller#7957] [PyInstaller] [answered]
Do you have any advice for my specific problem?

As a result of https://github.com/pyinstaller/pyinstaller/pull/7713, all included files are placed in an `_internal` directory which is placed alongside the binary. Because of this, I encountered a minor snag in my project's build script for Linux to convert a pyinstaller-generated onedir directory into an AppImage, as `appimagetool` expects there to be a desktop file and a icon file in the top-level directory as per the [AppImage Specification](https://docs.appimage.org/reference/appdir.html#id2).

This was relatively easy to fix by adding relative symlinks in the top-level directory which link to their counterpart in the `_internal` directory, as shown here:
https://github.com/zevlee/hello-world-gtk/blob/f206506d56272429bdb33400b3cf41e8f12d344c/linux/build.sh#L23-L24

However, it would be nice if I didn't need to do additional modification to the appdir prior to using `appimagetool`. Is there a recommended way for me to edit my spec file for this specific use case? It seems like https://github.com/pyinstaller/pyinstaller/pull/7619 is something I should learn more about, though I'm unsure if it's useful for what I'm trying to do.

**Comment by rokm** [comment, marked as answer]
The files that should not be relocated by PyInstaller to the `_internal` directory for whatever reason are strictly not part of PyInstaller-frozen application bundle; they are part of higher-level application bundle, and should be managed there. There was no such distinction before because the PyInstaller's top-level application directory (`sys._MEIPASS`) was the same directory that contained the executable (in onedir builds, at least; for onefile that was never the case).

So adding relative symbolic links like you did, or copying such files in the post-processing step (as opposed to having PyInstaller collect them) is the correct approach. https://github.com/pyinstaller/pyinstaller/pull/7619 is not relevant to your case, because all symlinking that PyInstaller is doing is supposed to be within its top-level application directory (so now within the `_internal`), not outside of it.

**Reply by zevlee** [reply]
I see, thank you for clarifying the scope of PyInstaller. Given that PyInstaller is already able to build macOS app bundles, I do wonder if it would be worthwhile to integrate AppImage functionality into PyInstaller. Such a project is something I would be interested in working on, if you would consider it worthwhile. What are your thoughts?

**Reply by rokm** [reply]
> Given that PyInstaller is already able to build macOS app bundles, I do wonder if it would be worthwhile to integrate AppImage functionality into PyInstaller. Such a project is something I would be interested in working on, if you would consider it worthwhile. What are your thoughts?

I think this is beyond the scope of PyInstaller itself, which I would very much prefer to keep limited to building frozen application bundles.

If you wanted to automate the AppImage build process, this would be better served as an external script/tool that calls PyInstaller to create frozen application, adjust the names, copies icon and desktop file, and calls the `appimagetool`. I.e., a higher-level tool that calls PyInstaller, rather than PyInstaller trying to orchestrate the AppImage assembly process. 

And since you mention macOS .app bundles; while we do assemble the .app bundle (due to specifics involved), we do not generate DMG files from them.
